### PR TITLE
Several bugs fixed

### PR DESCRIPTION
**Pull request by dvp2015**
Besides several bug fixes and PY3 incompatibilities I've added testing facility (pytest based). Without this the collective work (actually any work) is impossible.
The test passed both in PY2 and PY3. The code coverage 24% only, but I think it covers basic renumbering functionality.
To see coverage run `run_coverage` and see the results in the resulting htmlcov folder.

### CODE REVIEW

**Comment by travleev** [comment]
Thank you very much for the proposal! I will look at it (hopefully) at the weekend.

**Comment by dvp2015** [comment]
I've found one bug in setup.py and fixed it. The commit is in the same branch.

File: numjuggler/likefunc.py [hunk 1]
```diff
@@ -59,7 +59,7 @@ def __init__(self, log=False):
         self.ld = {}
 
         # Flag to store log
-        self.log = log
+        self.__lf = log
```
Comment (travleev) [inline numjuggler/likefunc.py:62]: In the original code, i.e. ``self.log = log`` the setter is called that sets the particular variant of the ``__call__`` method. In your case the ``__call2`` method will not be used when ``True`` is passed directly to the constructor.

File: numjuggler/main.py [hunk 2]
```diff
@@ -77,7 +86,7 @@ def tr2str(pl, fmt1='{:12.9f}', fmte='{:16.8e}'):
          'annotate', 'getc', 'mnew', 'combinec', 'cdens')
 
 
-def main():
+def main(args=sys.argv[1:]):
```
Comment (travleev) [inline numjuggler/main.py:89]: Why the optional argument ``args`` is needed here? It is later passed to the ``parse_known_args()`` method that refers anyway to ``sys.argv`` (according to [docs](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.parse_args)).

File: numjuggler/main.py [hunk 3]
```diff
@@ -260,7 +269,7 @@ def main():
             # as well.
 
             import re
-            r = re.compile('(u)(\d+)')
+            r = re.compile(r'(u)(\d+)')
```
Comment (travleev) [inline numjuggler/main.py:272]: Why this change? Both variants (with ``r`` and without) create the same pattern both in Python2 and Python3.

File: .coveragerc [hunk 4]
```diff
@@ -0,0 +1,33 @@
+[run]
```
Comment (travleev) [inline .coveragerc:1]: What is this file for? Could you please describe here as comments (or put links where to read)

File: .gitattributes [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#
```
Comment (travleev) [inline .gitattributes:1]: The repo contains no files satisfying the templates given in this file. Is this file really needed?

File: .pylintrc [hunk 6]
```diff
@@ -0,0 +1,407 @@
+[MASTER]
```
Comment (travleev) [inline .pylintrc:1]: This seems to be the pylint configuration file. I'm not familiar with pylint. Could you please describe why this file is needed in the repo?

File: numjuggler/nogq2.py [hunk 7]
```diff
@@ -1,444 +1,451 @@
...
-        cmnt.append('gamma_F = {:15.8e}'.format(g))
-
-    if areclose((0, D), atol=tDEF, rtol=None):
-        if areclose((1, B), atol=tABC, rtol=None):
-            g = 1.0/B
-            gammas.append(g)
-            cmnt.append('gamma_B = {:15.8e}'.format(g))
-        if areclose((1, A), atol=tABC, rtol=None):
-            g = 1.0/A
+from __future__ import print_function, division, nested_scopes
```
Comment (travleev) [inline numjuggler/nogq2.py:1]: ``git diff `` does not show properly the introduced changes. I suppose, the original code is wrapped with the ``if six.PY2 ...`` clause. This will result in preventing the ``--mode nogq2`` to work under Python3. Why?

File: numjuggler/parser.py [hunk 8]
```diff
@@ -19,21 +22,21 @@
     import pickle as cPickle
 
 # integer with one prefix character
-re_int = re.compile('\D{0,1}\d+')
+re_int = re.compile(r'\D{0,1}\d+')
```
Comment (travleev) [inline numjuggler/parser.py:25]: Here and in several other places: why all tepmlate strings start with ``r``? I tried both in Python2 and Python3: the re.compile() objects have the same patterns when defined with or without the ``r`` in front of the template string.

File: numjuggler/parser.py [hunk 9]
```diff
@@ -1150,44 +1155,54 @@ def is_blankline(l):
     """
     return l.strip() == ''
 
+if six.PY2:
```
Comment (travleev) [inline numjuggler/parser.py:1158]: Why the attempt to work with the dump file is wrapped with the ``if six.PY2 ...`` clause? Does it fails under Python3 or has no influence to the performance? (Under Python2 the use of dump file gives considerable gain in reading time, as tested on a ~ 40 MB input file)

File: numjuggler/utils/PartialFormatter.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
+from __future__ import print_function, absolute_import, division
```
Comment (travleev) [inline numjuggler/utils/PartialFormatter.py:1]: The ``PartialFormatter`` seems to provide some features to the underlying ``string.formatter``. COuld you please describe the features here, and why they are important in numjuggler?

File: numjuggler/utils/io.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
+import os
```
Comment (travleev) [inline numjuggler/utils/io.py:1]: Modules ``io.py`` and ``resource.py`` are used only for testing and seem to provide general functionality. If they are not numjuggler-specific, it is better to put them into a separate repository/package so they could be used e.g. in other repositories.

File: pytest.ini [hunk 12]
```diff
@@ -0,0 +1,14 @@
+[pytest]
```
Comment (travleev) [inline pytest.ini:1]: Could you provide a description of pytest (or put links where to read)? I think it would be helpful to put into a dedicated section in the repo's README information about:

- How to run pytest and how to integrate it into the common git work
- How to extend the set of test cases

**Review by travleev — COMMENTED** [review]
Once again: thank you very much for the contribution. 

Detailed comments/questions I placed as the comments to the files. Most of them are simply to understand why the changes were introduced. With one exception (``__lf`` insead of ``log``) the proposed changes can be accepted and merged. 

Concerning the Python3 support: I would like to separate the changes toward Python3 and put them e.g. into a dedicated branch. Originally numjuggler was developed and applied under Python2 only. I know only one user actively applying it in everyday work and he uses Python2. Do we really need to provide Python3 compatibility, at least right now? 

Regards, Anton.

File: numjuggler/main.py (hunk 2 above)
Comment (dvp2015) [reply numjuggler/main.py:89]: To facilitate tests with the main() function. Without this argts it would be necessary to change global variable sys.argv. Changing globals and even relying on them is not good idea: test cannot run in parallel, the design is not clear without reading docs and so on.

File: numjuggler/likefunc.py (hunk 1 above)
Comment (dvp2015) [reply numjuggler/likefunc.py:62]: Somewhere in the code there was a log call  to the attribute __lf, which appears to be bool value at the call moment. I've decide that this simple misprint and fixed this. Probably my fix doesn't cover __call2 functionality, but allowed to run renumbering.
We have to add test for logical branch where__call2 should be invoked.

File: numjuggler/main.py (hunk 3 above)
Comment (dvp2015) [reply numjuggler/main.py:272]: There was Python 3.7 warning: the pattern construct without 'r' became deprecated. Just to stop these warnings.

File: .coveragerc (hunk 4 above)
Comment (dvp2015) [reply .coveragerc:1]: To configure coverage. For example, don't include into the coverage lines like `__name__ == "__main__"`, which won't be passed in tests anyway. Try run_coverage and see the results in htmlcov folder, which will be created. Ideally the coverage should be 100%. As to me, I always (almost) follow test driven development methodology: start new function coding with tests, than add and develop the function itself. It's much more productive way for many reasons: you have to obey proper design, maintain integrity through any changes.  
Your code has grown up much above a use-once-and-throw-away utility and deserves this attitude.

File: numjuggler/likefunc.py (hunk 1 above)
Comment (travleev) [reply numjuggler/likefunc.py:62]: Thanks, indeed, there is the bug related to the ``log`` property`. I'll fix this directly on the master branch.

File: .coveragerc (hunk 4 above)
Comment (travleev) [reply .coveragerc:1]: Thanks. I suppose this is for [coverage](https://coverage.readthedocs.io/en/coverage-4.4.2/index.html) that one should run locally. Do I right understand, that here it is used to estimate the amount of code that is run during the tests?

File: numjuggler/main.py (hunk 2 above)
Comment (travleev) [reply numjuggler/main.py:89]: Ok, thanks.

File: numjuggler/main.py (hunk 3 above)
Comment (travleev) [reply numjuggler/main.py:272]: Ok, thanks.

File: .pylintrc (hunk 6 above)
Comment (dvp2015) [reply .pylintrc:1]: That's quite optional. Pylint is a static code analyzer. It polishes the code format and is able to find some mistakes. I just use it in all the python projects to check the files before pushing to public repos. I think it will be useful here too.

File: numjuggler/nogq2.py (hunk 7 above)
Comment (dvp2015) [reply numjuggler/nogq2.py:1]: The module relies on vector class from pirs. The pirs library is not PY3 compatible and I had some problems with installing it even in PY2 environment. I couldn't devote more time for this, so this is just quick fix. If you don't have pirs, you don't have nogq2.

By the way, why don't you use numpy as basis for this math? Numpy is nearly ubiquitous in python environments, so, it's not a heavy dependency.

File: numjuggler/parser.py (hunk 8 above)
Comment (dvp2015) [reply numjuggler/parser.py:25]: Looks like I've already answered this: to stop Python 3.7 deprecation warnings.

File: numjuggler/parser.py (hunk 9 above)
Comment (dvp2015) [reply numjuggler/parser.py:1158]: cPickle is not available in Python3, only the pickle module.Trying just to switch to pickle, I've met some issues with byte/unicode conversions and didn't have enough time to solve this. There's no performance issues for me, even with ITER models.
For caching I've used joblib. But these caches may be mysterious if not all dependencies are accounted on a cache update. I usually drop the caches every second run, when debugging the code.

File: numjuggler/utils/PartialFormatter.py (hunk 10 above)
Comment (dvp2015) [reply numjuggler/utils/PartialFormatter.py:1]: The MCNP models may contain entries like {31c}. The previous version of the code failed on this, looking for the kwargs dictionary a field with the name '31c'. The partial formatter ignores these entries. Otherwise you have to clean the original MCNP code from these entries, but this is not good idea: you'd better don't change the original files at all.

File: numjuggler/utils/io.py (hunk 11 above)
Comment (dvp2015) [reply numjuggler/utils/io.py:1]: These are just two files. Add repository just for them? Besides, to my experience latter these files become more specific for the project. These are just initial seeds.

File: pytest.ini (hunk 12 above)
Comment (dvp2015) [reply pytest.ini:1]: [pytest](https://docs.pytest.org/en/latest/) is probably the most popular python testing framework,
both easy to use and versatile. It's supported in all the IDE's I've used for Python so far: Eclipse, PyCharm, VSCode. The major scientific and math libraries Numpy, Scipy also use it.
You can add some test_smth.py file to the tests folder and add there some test_smth() method. Ordinary builtin asserts are redefined and produce information like much more complicated assertions from Unittests for example.

**Comment by dvp2015** [comment]
I think my version is somewhat better: you could share more information between the code and setup.py using `__version__.py` module.

File: numjuggler/likefunc.py (hunk 1 above)
Comment (dvp2015) [reply numjuggler/likefunc.py:62]: self.log=... creates attribute log
@property
def log()...         - creates method
both in the same class,  don't you think this is bad? These two guys must have different names to me.

Comment (travleev) [reply numjuggler/likefunc.py:62]: When ``def log(self)`` is [decorated](https://docs.python.org/2/library/functions.html#property) with ``@log.setter``, setting the respective attribute, i.e. ``self.log = value`` actually calls ``self.log(value)``.

File: .pylintrc (hunk 6 above)
Comment (travleev) [reply .pylintrc:1]: I think I use something like this, integrated into vim (it reports about syntax errors, formatting incompatible with pep8 etc.). I tried ``pylint main.py`` and it generated a huge list of messages: the resulting table reported about 375 invalid names. Do you use some command line parameters to control the output?

File: numjuggler/nogq2.py (hunk 7 above)
Comment (travleev) [reply numjuggler/nogq2.py:1]: Ok, agree.

> ... why don't you use numpy ...

Historically. For 5 years numpy was not so simple to install on a linux cluster into the user environment (now anaconda makes this simple).

File: numjuggler/utils/PartialFormatter.py (hunk 10 above)
Comment (travleev) [reply numjuggler/utils/PartialFormatter.py:1]: Just curiosity: the entry like ``{31c}`` is a valid MCNP input syntax, or occasionally can be found within the comments?

File: pytest_coverage [hunk 13]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+
+pytest --cov-report=term-missing:skip-covered --cov=numjuggler $*
```
Comment (travleev) [inline pytest_coverage:3]: This script generated an error:

Under Python2:

<pre>(base) <font color="#AF0000">anton@local-Laptop1806 numjuggler&gt; </font>./pytest_coverage 
usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --cov-report=term-missing:skip-covered --cov=numjuggler
  inifile: /home/anton/work/gh/numjuggler/pytest.ini
  rootdir: /home/anton/work/gh/numjuggler
(base) <font color="#AF0000">anton@local-Laptop1806 numjuggler&gt; </font>pytest --version
This is pytest version 4.0.2, imported from /home/anton/anaconda2/lib/python2.7/site-packages/pytest.pyc
(base) <font color="#AF0000">anton@local-Laptop1806 numjuggler&gt; </font>
</pre>

Under Python3:

<pre>(base) <font color="#AF0000">anton@local-Laptop1806 numjuggler&gt; </font>./pytest_coverage 
usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --cov-report=term-missing:skip-covered --cov=numjuggler
  inifile: /home/anton/work/gh/numjuggler/pytest.ini
  rootdir: /home/anton/work/gh/numjuggler
(base) <font color="#AF0000">anton@local-Laptop1806 numjuggler&gt; </font>pytest --version
This is pytest version 4.0.2, imported from /home/anton/anaconda3/lib/python3.7/site-packages/pytest.py
setuptools registered plugins:
  pytest-remotedata-0.3.1 at /home/anton/anaconda3/lib/python3.7/site-packages/pytest_remotedata/plugin.py
  pytest-openfiles-0.3.1 at /home/anton/anaconda3/lib/python3.7/site-packages/pytest_openfiles/plugin.py
  pytest-doctestplus-0.2.0 at /home/anton/anaconda3/lib/python3.7/site-packages/pytest_doctestplus/plugin.py
  pytest-arraydiff-0.3 at /home/anton/anaconda3/lib/python3.7/site-packages/pytest_arraydiff/plugin.py
(base) <font color="#AF0000">anton@local-Laptop1806 numjuggler&gt; </font>
</pre>

File: .coveragerc (hunk 4 above)
Comment (dvp2015) [reply .coveragerc:1]: After running coverage you could see the complete report (by default, in htmlcov directory) on the lines of the code visited on test runs. Yes, it also reports percentage of code coverage and range of lines missed on console.

File: .pylintrc (hunk 6 above)
Comment (dvp2015) [reply .pylintrc:1]: There should be some options, you can add pylint ignore statements as special comments in the code. I've never met that many invalid names. I'm just forced to follow the naming rules from the very beginning. There are some exclusions In pylintrc, you can add their yours.

File: numjuggler/utils/PartialFormatter.py (hunk 10 above)
Comment (dvp2015) [reply numjuggler/utils/PartialFormatter.py:1]: This entry happens in comments

File: numjuggler/utils/io.py (hunk 11 above)
Comment (dvp2015) [reply numjuggler/utils/io.py:1]: Just recently I've worked with likefunction code and added tests for it. To do that, I've extended a bit and moved some common code to the io.py . So, it's growing and, certainly, will grow further.
These changes will appear with the next change request

File: numjuggler/likefunc.py (hunk 1 above)
Comment (dvp2015) [reply numjuggler/likefunc.py:62]: Looks like all this is fixed in the next pull request

File: .gitattributes (hunk 5 above)
Comment (dvp2015) [reply .gitattributes:1]: It's not harmful. What if someone comes up with a solution using zipped data resources for example and forget to create the file? Or what if someone adds a Word document to justify or explain the task?
The Git would better to know which files are binary. I usually save this file in all my projects.
If you don't like it, let's remove until it needed.

File: pytest_coverage (hunk 13 above)
Comment (dvp2015) [reply pytest_coverage:3]: You need pytest-cov module installed. The module is listed in setup.py task_require parameter (in my version).

Comment (dvp2015) [reply pytest_coverage:3]: On Python2 or Python3 see [this](https://pythonclock.org/) and [this](https://www.anaconda.com/end-of-life-eol-for-python-2-7-is-coming-are-you-ready/)
